Add is_subset handler for Range in FiniteSet

gschintgen opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests . Please also
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #18124.
Fixes #18146.

#### Brief description of what is fixed or changed
Add `is_subset` handler for (Range, FiniteSet) and tests.
Fixes logic errors in `Range.contains()`

#### Other comments
Two missed lines (marked red in codecov.io) in sets/handlers/issubset.py should now also be covered.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* sets
  * Added an `is_subset` handler to check whether a `Range` is included in a `FiniteSet`.
  * Added a simple cache to `FiniteSet` in order to speed up some membership checks.
  * Fixed `Range.contains()` logic for symbolic values.
<!-- END RELEASE NOTES -->

gschintgen committed:
Add is_subset handler for (Range, FiniteSet) and tests

gschintgen committed:
Improve coverage of is_subset (slightly)

oscarbenjamin commented on sympy/sets/handlers/issubset.py:
```diff
@@ -66,6 +66,18 @@ def is_subset_sets(a_range, b_interval): # noqa:F811
+    if a_size > len(b_finiteset):
+        return False
+    else:
+        return fuzzy_and(b_finiteset.contains(x) for x in a_range)
```
I guess we could use this logic for any finite iterable set rather than just `Range`. It could be slow though e.g.:
```
In [5]: S = FiniteSet(*(Dummy() for _ in range(10**3)))

In [6]: Range(10**3).is_subset(S) # Really slow...
```

czgdp1807 replied:
May be this can be analysed just by using endpoints of `FiniteSet` and `Range`. @gschintgen Have you given this idea a consideration?

gschintgen replied:
> Range(10**3).is_subset(S) # Really slow...

That's a pity. Even a *single* `S.contains(5)` takes almost 2 seconds (on a slow system). I concede that I must have imagined `FiniteSet` to be hash-based so that a contains-check would be almost instant (and constant-time relative to the cardinality). But in hindsight it's obvious that, in order to give mathematically meaningful answers, `contains` must indeed check equality with every element of the `FiniteSet`. So it comes down to 2ms per equality check which sums up rather quickly. Hmm. (2ms is still expensive for a rather basic check, isn't it?)
So the problem with the proposed code is that its running time scales with the product of the cardinalities of both sets and that's not really acceptable.

@czgdp1807 : I'm not sure I completely understand your idea. A FiniteSet may contain an arbitrary number of arbitrary expressions. What could potentially be done is somehow filtering/pre-processing the FiniteSet in a single run-through (e.g. kick out irrelevant non-integers). I'll think about it.

gschintgen replied:
Oh well, this took me some time. I've added a new commit that changes the logic. I just hope I got my testing right. Please review, but note the gotcha below that I discovered only very late...

With this new commit, the example given by @oscarbenjamin is handled almost instantly by the handler, **but** there's another **huge** slowdown in line 385:
https://github.com/sympy/sympy/blob/f7390e5fa2313be1c49959270358b2b0c62a1a57/sympy/sets/sets.py#L377-L386
The `None` case is already pathologically slow in current master due to the intersection computation.

I also note that the logic of that intersection test is bound to fail for Range.intersect(FiniteSet): The result will be a FiniteSet and hence test unequal to self. So the expensive computation is in vain.

So to sum up (for the slow example):
master: slow due to useless intersection computation
first commit: slow handler, followed by slow intersection
new commit: faster handler, followed by slow intersection.

oscarbenjamin replied:
Not sure if I added that comment but it sounds like something I'd say...

oscarbenjamin replied:
I wonder if it's possible to speed up general membership testing with something like this. Maybe the `FiniteSet` could cache a `set` that can be used for fast membership testing as separate from the more complicated symbolic cases.

gschintgen replied:
I've added a basic cache. (In fact a temporary `set` of all elements/args was created anyway in `__new__`, so it was sufficient to keep it around.) For larger `FiniteSets` it's indeed beneficial to first check membership using this cache. But only the `True` case with an exact match is sped up.
(See commit message for a cursory benchmark.)

oscarbenjamin replied:
It doesn't need to be done in this PR but I guess there are some situations where we know based on type that `==` is sufficient and therefore can use efficient set algorithms. For example if all elements of both sets are of type `Number` then we can process everything using `set` because we know that `Eq(a, b)` is True if `a==b` and False otherwise (not sure about `Float`...). On other hand if any element is a symbol with no defined assumptions then we know that `Eq(a, b)` will be True if `a == b` and "None" otherwise so probably something can be done with that as well.

oscarbenjamin replied:
I think we've probably over-complicated things here. Presumably the simple contains loop would cover the case
```
Range(n, n + 2).is_subset(FiniteSet(n, n + 1))
```
which with the PR now is None rather than True.

I guess we should at least fall back on the slow method here. It's probably more important to do the right thing with small symbolic sets than worry about performance with larger sets.

gschintgen replied:
> ```
> Range(n, n + 2).is_subset(FiniteSet(n, n + 1))
> ```
> 
> 
> which with the PR now is None rather than True.

I agree that for a CAS symbolics should be more important than pure performance. I'm somewhat astonished that in the early commit with the basic `fuzzy_and` check I included this test with `None` as expected result. (The code of the very first commit indeed returns `True`!) I know I did some trials with symbolic `Ranges` but must have misinterpreted something. Possibly a mishap with the `n` in the tests file that doesn't have integer assumptions combined with #18146 cropping up. I think I even concluded that `Range` and symbols don't mix well...

I'll see what I can do.

gschintgen replied:
@oscarbenjamin, @czgdp1807 : What do you think now with the latest commits?

As for the more advanced caching, it's an interesting idea and I'm sure there's more that could be done (e.g. not only check if all elements are of certain types, but maybe cache precisely which `args` are of other/symbolic types, so that only those specific elements need to be checked.) But at this stage I'm not sure how important this optimization for large `FiniteSets` is. I'd rather weed out logic errors, sets-related or not.

oscarbenjamin replied:
> I'd rather weed out logic errors, sets-related or not.

Agreed

gschintgen committed:
Rewrite is_subset handler for (Range, FiniteSet)

It turns out that membership checks on a large FiniteSet can be
quite slow. (Runtime grows linearly with cardinality.)
The new logic introduced by this commit converts the Range
to a plain Python set in order to take advantage of its hashing.

gschintgen committed:
Update FiniteSet._contains() docstring to reflect current code

gschintgen committed:
Use ASCII-friendly spelling 'naive'

gschintgen committed:
Add _args_set cache to FiniteSet in order to speed up `._contains`

Before:

In [12]: f = FiniteSet(*range(10**4))

In [13]: timeit(f.contains(9999))
1 loop, best of 3: 2 s per loop

After:

In [1]: f = FiniteSet(*range(10**4))

In [2]: timeit(f.contains(9999))
The slowest run took 8.01 times longer than the fastest. This could mean
that an intermediate result is being cached.
100000 loops, best of 3: 4.08 µs per loop

gschintgen committed:
Re-add support for symbolic Ranges in is_subset handler

gschintgen committed:
Fix Range.contains(n) for symbolic n

Fixes #18146.
(The results are mathematically valid, but not optimal if the
Range is reduced to a single element.)

A number of tests are added.

gschintgen committed:
Simplify relational for single element Ranges

oscarbenjamin commented:
Looks good. Merging.
